### ISSUE TITLE
Comms vat failure against dapp-token-economy

### ISSUE THREAD

**Issue by dtribble** [Agoric/agoric-sdk#2588]
## Describe the bug

When deploying the token economy contracts, I get a comms failure: `Error#5: clist violation: prepare to die `

Full report:
https://gist.github.com/dtribble/20f16d9105a2ac820792069f4d3837c6

## To Reproduce
1. using agoric-sdk `master` at caa367d9ca355feb82b79928cde8eb92b4c093bf
1. as usual, `yarn install` and `yarn build`
2. in `dapp-token-economy`, run `agoric start -v --reset`
2. in `dapp-token-economy` with `vaultManager:57` changed to use `E(timerService).` rather than `timerService.`
3. `agoric deploy contract/deploy.js`
4. in the wallet, enable `TreasuryDeploy` (do `agoric open` to bring up browser)
5. `agoric deploy api/deploy.js`

## Expected behavior
The deploy operation should complete

## Platform Environment
 - WSL with node 14.16.0
 - one change to allow the deploy to proceed: 
    - `vaultManager:57` changed to use `E(timerService).` rather than `timerService.`
 - what version of the Agoric-SDK are you using? (run `git describe --tags --always`)

## Additional context
Full data here: https://gist.github.com/dtribble/20f16d9105a2ac820792069f4d3837c6

I tried a variety of nearby branches. The only way I could proceed was to comment out the timerService usage for a notifier:

```javascript
  const periodNotifier = E(timerService).makeNotifier(
    0n,
    loanParams.recordingPeriod,
  );
  const { zcfSeat: poolIncrementSeat } = zcf.makeEmptySeatKit();
  observeIteration(periodNotifier, {
    updateState: updateTime => chargeAllVaults(updateTime, poolIncrementSeat),
  });
```
That resolves promises through the comm system.

**Comment by warner** [comment]
Hm, given that the fatal syscall happened during an *un*serialize:

```
2021-03-06T19:13:33.056Z SwingSet: kernel: Temporary logging of sent error (Error#5)
Error#5: clist violation: prepare to die
  at fullRevive (packages/marshal/src/marshal.js:966:36)
  at fullRevive (packages/marshal/src/marshal.js:1009:23)
  at meteredConstructor.unserialize (packages/marshal/src/marshal.js:1049:19)
  at meteredConstructor.deliver (packages/SwingSet/src/kernel/liveSlots.js:462:20)
  at eval (packages/SwingSet/src/kernel/vatManager/deliver.js:51:48)
```

(https://gist.github.com/dtribble/20f16d9105a2ac820792069f4d3837c6#file-gistfile1-txt-L282-L288)

I wonder if `convertSlotToVal` encountered an ancillary/ephemeral Promise ID and did a `syscall.subscribe()`, when in fact the kernel had a resolution being delivered in the same batch, so liveslots should have deferred doing the subscribe until it saw that the new promise was left unresolved after finishing the complete batch. (This assumes the `unserialize` happened in a `dispatch.notify` and not a `dispatch.deliver`.. the stack trace shows a function named `deliver` but I'm not currently going to assume that it's the `dispatch` method until looking at the code).

**Comment by erights** [comment]
The line at marshal.js:966 is
```js
            const error = harden(new EC(`${rawTree.message}`));
```
This looks like the error did not happen *during* unserialize, but rather that such an error was serialized and was here being unserialized. However, I'm not sure. If that was the case, we should also see an 
```
Error#5 ERROR_NOTE: Received as ...
```
which does not appear in that gist. Taken together, I don't understand the symptoms I see.

**Comment by FUDCo** [comment]
@erights is correct that the error didn't happen during unserialize, that's just where the error itself was subject to the scrutiny of `marshal`.  It's happening somewhere in one of the ZCF vats, I think on a `resolve` syscall.  In particular, it really doesn't look to me like the comms vat is implicated in this, but I won't know for sure without a bit more detail on the reference that it was unhappy with.  I think the error itself could benefit from being more descriptive about that, which I'll change forthwith.

**Comment by FUDCo** [comment]
Nope, the problem was not caused by a vpid resurrection being passed to `resolve`.  Running in verbose mode revealed that the underlying diagnostic says `vat v16 terminated: error during translation: TypeError: Cannot convert a Symbol value to a string ["send","p+141",null,{"body":"[]","slots":[]},"p+143"].`  I strongly suspect that the `null` in the message is not a real null but what `JSON.stringify` (called in creating the error string) produced when given a Symbol. Probably that error should use `q` instead.  The issue manifested as a clist violation because the code that caught the exception was written assuming that the only thing that could go wrong at that point was a clist problem, but that's not what happened.  Obviously there's an issue making the error message more correct, but ultimately the problem is that somebody tried to send a message containing a Symbol.  This also raises the question of what the appropriate action is if somebody does this; killing the vat feels to me like a denial of service opportunity, but if we take the position that code inside a vat always can misbehave in ways that kill the vat then we would be OK just fixing the error message.  I have a dim memory we've discussed this kind of thing before and came up with a general policy position, but if so I don't remember what it was.

**Comment by FUDCo** [comment]
Further investigation reveals that the problem is that somebody is using a Symbol, specifically, `Symbol.asyncIterator`, as a method name.  This simply does not work, as symbols as method names have not been plumbed through the kernel.

**Comment by dtribble** [comment]
So inside the pattern of using `observeIteration` in the OP came from @erights, what should this code be doing instead?  It is inside the contract. I think this id going from the `timerService` to the treasury, which just reacts by calling a local function to update state.

**Comment by erights** [comment]
> Use `observeNotifier` instead.

Only if what's being observed is a notifier. Just noting that we still need https://github.com/Agoric/agoric-sdk/issues/2092 fixed so we can observe iterations that are not notifiers.

**Comment by erights** [comment]
> So inside the pattern of using `observeIteration` in the OP came from @erights

@dtribble ,
"OP"?

**Comment by FUDCo** [comment]
#2481 should take care of this

**Comment by warner** [comment]
We addressed this with #2619 , which special-cases `Symbol.asyncIterator` (translating the Symbol method name into the literal string `'Symbol.asyncIterator'`) to allow this pattern to work despite SwingSet not supporting Symbol-named methods in general.

#2481 is a more general fix, that we aren't doing this week.
